src/main/java/io/ballerina/plugins/idea/editor/BallerinaFoldingBuilder.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.com).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.ballerina.plugins.idea.editor;
+
+import com.intellij.lang.ASTNode;
+import com.intellij.lang.folding.CustomFoldingBuilder;
+import com.intellij.lang.folding.FoldingDescriptor;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.project.DumbAware;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiComment;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiWhiteSpace;
+import com.intellij.psi.tree.IElementType;
+import com.intellij.psi.util.PsiTreeUtil;
+import io.ballerina.plugins.idea.psi.BallerinaDocumentationString;
+import io.ballerina.plugins.idea.psi.BallerinaFile;
+import io.ballerina.plugins.idea.psi.BallerinaImportDecl;
+import io.ballerina.plugins.idea.psi.BallerinaTypes;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Stack;
+
+/**
+ * Folding builder for Ballerina language.
+ *
+ * @since 2.0.0
+ */
+public class BallerinaFoldingBuilder extends CustomFoldingBuilder implements DumbAware {
+
+    private static final String DEFAULT_PLACEHOLDER = "...";
+    private static final String COMMENT_PLACEHOLDER = "// ...";
+    private static final String DOCUMENTATION_PLACEHOLDER = "# ...";
+    private static final String BRACE_PLACEHOLDER = "{...}";
+    private static final String BRACE_PIPE_PLACEHOLDER = "{|...|}";
+
+    @Override
+    protected void buildLanguageFoldRegions(@NotNull List<FoldingDescriptor> list, @NotNull PsiElement psiElement,
+                                            @NotNull Document document, boolean b) {
+        if (!(psiElement instanceof BallerinaFile)) {
+            return;
+        }
+        matchPair(list, psiElement, BallerinaTypes.OPEN_BRACE_TOKEN, BallerinaTypes.CLOSE_BRACE_TOKEN,
+                BRACE_PLACEHOLDER);
```

Comment:
I prefer the following pattern for creating and adding folding regions to the existing list

```suggestion
        list.addAll(createFoldingRegions(psiElement, BallerinaTypes.OPEN_BRACE_TOKEN, BallerinaTypes.CLOSE_BRACE_TOKEN,
                BRACE_PLACEHOLDER));
```